### PR TITLE
[WIP] Change plasma_get take a timeout and multiple object IDs.

### PR DESCRIPTION
**Pull request by robertnishihara**
Changes:
- `plasma_get` now takes a timeout.
- `plasma_get` takes an array of object IDs.
- fixed a bug in the implementation of `plasma_wait` #208.
- removed old code

TODO:
- add more tests for new functionality.

### CODE REVIEW

File: src/plasma/plasma_store.c [hunk 1]
```diff
@@ -208,42 +234,232 @@ int create_object(client *client_context,
...
+  remove_get_request(store_state, get_req);
+}
+
+void update_object_get_requests(plasma_store_state *store_state,
+                                object_id obj_id) {
+  /* Update the in-progress get requests. */
+  object_get_requests *object_get_reqs;
+  HASH_FIND(hh, store_state->object_get_requests, &obj_id, sizeof(obj_id),
+            object_get_reqs);
+  if (object_get_reqs != NULL) {
```
Comment (stephanie-wang) [inline src/plasma/plasma_store.c]: We could remove one indentation by returning if `object_get_reqs == NULL`.

File: src/plasma/plasma_store.c [hunk 2]
```diff
@@ -208,42 +234,232 @@ int create_object(client *client_context,
...
+       * object is not present. This will be parsed by the client. */
+      get_req->objects[i].data_size = -1;
+      /* Add the get request to the relevant data structures. */
+      add_get_request_for_object(plasma_state, obj_id, get_req);
     }
-    utarray_push_back(notify_entry->waiting_clients, &client_context);
   }
-  return OBJECT_NOT_FOUND;
+
+  /* If all of the objects are present already, return to the client. */
```
Comment (stephanie-wang) [inline src/plasma/plasma_store.c]: Might be worth it to also return immediately if `timeout_ms` is 0, since this will be a common pattern.

**Review by stephanie-wang — CHANGES_REQUESTED** [review]

File: src/plasma/plasma_store.c (hunk 2 above)
Comment (robertnishihara) [inline src/plasma/plasma_store.c]: Good idea.

File: lib/python/ray/worker.py [hunk 3]
```diff
@@ -428,20 +428,25 @@ def put_object(self, objectid, value):
     # Optionally do something with the contained_objectids here.
     contained_objectids = []
 
-  def get_object(self, objectid):
+  def get_object(self, object_ids):
```
Comment (pcmoritz) [inline lib/python/ray/worker.py]: can you rename this to get_objects and fix the docstring?

File: src/plasma/plasma_client.c [hunk 4]
```diff
@@ -270,55 +270,93 @@ int plasma_create(plasma_connection *conn,
...
+
+  /* If all of the objects are currently in use by the client, we can avoid
+   * sending a message to the store. */
+  if (all_present) {
+    for (int i = 0; i < num_objects; ++i) {
+      object_in_use_entry *object_entry;
+      HASH_FIND(hh, conn->objects_in_use, &object_ids[i], sizeof(object_ids[i]),
+                object_entry);
+      DCHECK(object_entry != NULL);
+      plasma_object object_data;
```
Comment (robertnishihara) [inline src/plasma/plasma_client.c]: Ok, good point.

**Comment by robertnishihara** [comment]
Replaced by #212.

**Pull request closed without merging**